src/main/drivers/bus_spi_hal.c

```diff
@@ -0,0 +1,263 @@
+    spi->hspi.Init.Mode = SPI_MODE_MASTER;
+    spi->hspi.Init.Direction = SPI_DIRECTION_2LINES;
+    spi->hspi.Init.DataSize = SPI_DATASIZE_8BIT;
+    spi->hspi.Init.NSS = SPI_NSS_SOFT;
+    spi->hspi.Init.FirstBit = SPI_FIRSTBIT_MSB;
+    spi->hspi.Init.CRCPolynomial = 7;
+    spi->hspi.Init.BaudRatePrescaler = SPI_BAUDRATEPRESCALER_256;
+    spi->hspi.Init.CRCCalculation = SPI_CRCCALCULATION_DISABLE;
+    spi->hspi.Init.TIMode = SPI_TIMODE_DISABLED;
+    spi->hspi.Init.FifoThreshold = SPI_FIFO_THRESHOLD_01DATA;
+    spi->hspi.Init.MasterKeepIOState = SPI_MASTER_KEEP_IO_STATE_ENABLE;  /* Recommanded setting to avoid glitches */
+
+    if (spi->leadingEdge) {
+        spi->hspi.Init.CLKPolarity = SPI_POLARITY_LOW;
+        spi->hspi.Init.CLKPhase = SPI_PHASE_1EDGE;
+    }
+    else {
+        spi->hspi.Init.CLKPolarity = SPI_POLARITY_HIGH;
+        spi->hspi.Init.CLKPhase = SPI_PHASE_2EDGE;
+    }
+
+    // Init SPI hardware
+    HAL_SPI_Init(&spi->hspi);
+}
+
+// return uint8_t value or -1 when failure
+uint8_t spiTransferByte(SPI_TypeDef *instance, uint8_t out)
+{
+    uint8_t in;
+
+    spiTransfer(instance, &out, &in, 1);
+    return in;
+}
+
+/**
+ * Return true if the bus is currently in the middle of a transmission.
+ */
+bool spiIsBusBusy(SPI_TypeDef *instance)
+{
+    SPIDevice device = spiDeviceByInstance(instance);
+    if(spiDevice[device].hspi.State == HAL_SPI_STATE_BUSY)
+        return true;
+    else
+        return false;
+}
+
+bool spiTransfer(SPI_TypeDef *instance, const uint8_t *out, uint8_t *in, int len)
+{
+    SPIDevice device = spiDeviceByInstance(instance);
+    HAL_StatusTypeDef status;
+
+#define SPI_DEFAULT_TIMEOUT 10
+
+    if (!in) {
+        // Tx only
+        status = HAL_SPI_Transmit(&spiDevice[device].hspi, out, len, SPI_DEFAULT_TIMEOUT);
+    } else if(!out) {
+        // Rx only
+        status = HAL_SPI_Receive(&spiDevice[device].hspi, in, len, SPI_DEFAULT_TIMEOUT);
+    } else {
+        // Tx and Rx
+        status = HAL_SPI_TransmitReceive(&spiDevice[device].hspi, out, in, len, SPI_DEFAULT_TIMEOUT);
+    }
+
+    if(status != HAL_OK) {
+        spiTimeoutUserCallback(instance);
+    }
+
+    return true;
+}
+
+// Position of Prescaler bits are different from MCU to MCU
+
+static uint32_t baudRatePrescaler[8] = {
+    SPI_BAUDRATEPRESCALER_2,
+    SPI_BAUDRATEPRESCALER_4,
+    SPI_BAUDRATEPRESCALER_8,
+    SPI_BAUDRATEPRESCALER_16,
+    SPI_BAUDRATEPRESCALER_32,
+    SPI_BAUDRATEPRESCALER_64,
+    SPI_BAUDRATEPRESCALER_128,
+    SPI_BAUDRATEPRESCALER_256,
+};
+
+void spiSetDivisor(SPI_TypeDef *instance, uint16_t divisor)
+{
+    SPIDevice device = spiDeviceByInstance(instance);
+
+    HAL_SPI_DeInit(&spiDevice[device].hspi);
+
+    spiDevice_t *spi = &(spiDevice[device]);
+
+    int prescalerIndex = ffs(divisor) - 1;
+
+    if (prescalerIndex < 0 || prescalerIndex >= (int)ARRAYLEN(baudRatePrescaler)) {
+        return;
+    }
+
+    spi->hspi.Init.BaudRatePrescaler = baudRatePrescaler[prescalerIndex - 1];
```

Comment:
prescaller index is checked `<0`, but `prescalerIndex - 1` is used here - passing divisor = 1 will cause underflow

Reply:
This is still prone to underflow if `prescalerIndex == 0`. Is this intentional?